File: server/src/database/dynamodb-handler.ts
import AWS from "aws-sdk"
import {DatabaseHandler} from "./database-handler";
import {MatchRecord, PlayerRecord} from "./records";
import {Match} from "../match";

const config = require("../pingu-config.json");

export class DynamodbHandler implements DatabaseHandler {

    private static PLAYER_TABLE_NAME: string =  "PinguPlayers";
    private static MATCH_TABLE_NAME: string =  "PinguMatches";

    private mDatabase: AWS.DynamoDB.DocumentClient;
    private mHighestPlayerIndex: number = -1;

    constructor() {
        AWS.config.update({
            region: config.awsRegion
        });
        this.mDatabase = new AWS.DynamoDB.DocumentClient();

        this.fetchPlayers().then((records) => {
            records.forEach((record) => {
               if (record.id > this.mHighestPlayerIndex) {
                   this.mHighestPlayerIndex = record.id;
               }
            });
        });
    }

    public async fetchPlayers(): Promise<PlayerRecord[]> {
        return new Promise((resolve, reject) => {
            this.mDatabase.scan({TableName: DynamodbHandler.PLAYER_TABLE_NAME}, (error, data) => {
                if (error) {
                    console.error(error.message);
                    reject(error);
                } else {
                    resolve(data.Items.map((item) => {
                        return {
                            id: item.id,
                            name: item.name
                        }
                    }));
                }
            });
        });
    }

    fetchMatches(): Promise<MatchRecord[]> {
        return new Promise((resolve, reject) => {
            this.mDatabase.scan({TableName: DynamodbHandler.MATCH_TABLE_NAME}, (error, data) => {
                if (error) {
                    console.error(error.message);
                    reject(error);
                } else {
                    resolve(data.Items.map((item) => {
                        return {
                            date: item.date,
                            winner: item.winner,
                            loser: item.loser,
                            winner_score: item.winner_score,
                            loser_score: item.loser_score,
                        }
                    }));
                }
            });
        });
    }

    fetchPlayerFromId(player: number): Promise<PlayerRecord> {
        return new Promise((resolve, reject) => {
            this.mDatabase.get({TableName: DynamodbHandler.PLAYER_TABLE_NAME, Key: {id: player}}, (error, data) => {
                if (error) {
                    console.error(error.message);
                    reject(error);
                } else {
                    resolve({
                        id: data.Item.id,
                        name: data.Item.name,
                    });
                }
            });
        });
    }

    createPlayer(name: string): Promise<void> {
        const data = {
            TableName: DynamodbHandler.PLAYER_TABLE_NAME,
            Item: {
                "id": this.mHighestPlayerIndex + 1,
                "name": name
            }
        };
        return new Promise((resolve, reject) => {
            this.mDatabase.put(data, (error, data) => {
                if (error != undefined) {
                    console.error(error);
                    reject(error);
                } else {
                    this.mHighestPlayerIndex++;
                    resolve();
                }
            })
        });
    }

    recordMatch(match: Match): Promise<void> {
        const data = {
            TableName: DynamodbHandler.MATCH_TABLE_NAME,
            Item: {
                "date": match.date,
                "winner": match.winner,
                "loser": match.loser,
                "winner_score": match.winnerScore,
                "loser_score": match.loserScore
            }
        };
        return new Promise((resolve, reject) => {
            this.mDatabase.put(data, (error, data) => {
                if (error != undefined) {
                    console.error(error);
                    reject(error);
                } else {
                    resolve();
                }
            })
        });
    }
}

File: kotlin-server/src/main/kotlin/com/example/kotlinserver/database/DynamoDbHandler.kt
package com.example.kotlinserver.database

import com.example.kotlinserver.Configuration
import com.example.kotlinserver.models.MatchModel
import com.example.kotlinserver.models.PlayerModel
import software.amazon.awssdk.auth.credentials.InstanceProfileCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.dynamodb.DynamoDbClient
import software.amazon.awssdk.services.dynamodb.model.AttributeValue
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest
import software.amazon.awssdk.services.dynamodb.model.ScanRequest
import java.time.LocalDate
import java.time.format.DateTimeFormatter

private const val PLAYERS_TABLE_NAME = "PinguPlayers"
private const val MATCH_TABLE_NAME = "PinguMatches"

class DynamoDbHandler : DatabaseHandler {

    private val dynamoDbClient: DynamoDbClient
    private var highestPlayerIndex = -1

    init {
        val provider = InstanceProfileCredentialsProvider.create()
        dynamoDbClient = DynamoDbClient.builder().region(Region.of(Configuration.awsRegion)).credentialsProvider(provider).build()

        fetchPlayers().forEach {
            if (it.id > highestPlayerIndex) {
                highestPlayerIndex = it.id
            }
        }
    }

    override fun fetchPlayers(): List<PlayerModel> {

        val request = ScanRequest.builder().tableName(PLAYERS_TABLE_NAME).build()
        val response = dynamoDbClient.scan(request)

        val players: MutableList<PlayerModel> = mutableListOf()

        response.items().toList().forEach { entries ->
            players.add(PlayerModel(
                DynamoDbParser.parseNumberAttribute(entries["id"].toString()),
                DynamoDbParser.parseStringAttribute(entries["name"].toString())
            ))
        }

        return players
    }

    override fun fetchMatches(from: LocalDate, to: LocalDate): List<MatchModel> {
        val request = ScanRequest.builder().tableName(MATCH_TABLE_NAME).build()
        val response = dynamoDbClient.scan(request)

        val matches: MutableList<MatchModel> = mutableListOf()

        response.items().toList().forEach { entries ->
            matches.add(MatchModel(
                DynamoDbParser.parseStringAttribute(entries["date"].toString()),
                DynamoDbParser.parseNumberAttribute(entries["winner"].toString()),
                DynamoDbParser.parseNumberAttribute(entries["loser"].toString()),
                DynamoDbParser.parseNumberAttribute(entries["winner_score"].toString()),
                DynamoDbParser.parseNumberAttribute(entries["loser_score"].toString()),
            ))
        }

        return matches
            .filter { LocalDate.parse(it.date, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")).isAfter(from) }
            .filter { LocalDate.parse(it.date, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")).isBefore(to) }
            .sortedByDescending { it.date }
    }

    override fun fetchPlayerFromId(playerId: Int): PlayerModel {
        return fetchPlayers().first { it.id == playerId }
    }

    override fun createPlayer(name: String) {
        highestPlayerIndex++
        val attributes = mapOf(
            Pair("id", numberAttributeValue(highestPlayerIndex)),
            Pair("name", stringAttributeValue(name)),
        )

        val request = PutItemRequest.builder().tableName(PLAYERS_TABLE_NAME).item(attributes).build()
        dynamoDbClient.putItem(request)
    }

    override fun recordMatch(match: MatchModel) {
        val attributes = mapOf(
            Pair("date", stringAttributeValue(match.date)),
            Pair("winner", numberAttributeValue(match.winner)),
            Pair("loser", numberAttributeValue(match.loser)),
            Pair("winner_score", numberAttributeValue(match.winnerScore)),
            Pair("loser_score", numberAttributeValue(match.loserScore)),
        )

        val request = PutItemRequest.builder().tableName(MATCH_TABLE_NAME).item(attributes).build()
        dynamoDbClient.putItem(request)
    }

    private fun numberAttributeValue(number: Int): AttributeValue {
        return AttributeValue.fromN(number.toString())
    }
    
    private fun stringAttributeValue(string: String): AttributeValue {
        return AttributeValue.fromS(string)
    }
}